### PR TITLE
Support all Spark patterns in cast(varchar as date)

### PR DESCRIPTION
**Pull request by marin-ma**
Below patterns are considered valid to cast from string to date in spark sql functions:

1. Year only: "YYYY"
2. Year and month only: "YYYY-MM"
3. Any characters after trailing spaces: "YYYY-MM-DD ", "YYYY-MM-DD 123", "YYYY-MM-DD (BC)"
4. Any characters after trailing character 'T': "YYYY-MM-DDT"

Below patterns are invalid:

1. Year is too large (exceed INT32_MAX): 20150318
2. Other separators "YYYY/MM/DD"

Reference: 
Spark cast from string to date:
https://github.com/apache/spark/blob/3e5203c64c06cc8a8560dfa0fb6f52e74589b583/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L286-L298

Unit test:
CastExprTest::fromStringToDate is derived from https://github.com/apache/spark/blob/3a9185964a0de3c720a6b77d38a446258b73468e/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala#L103-L126
CastExprTest::fromStringToDateInvalid is derived from 
https://github.com/apache/spark/blob/3a9185964a0de3c720a6b77d38a446258b73468e/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastWithAnsiOffSuite.scala#L67-L73

### CODE REVIEW

**Comment by marin-ma** [comment]
@mbasmanova PTAL. Thanks!

**Review by mbasmanova — COMMENTED** [review]
@marin-ma Does Presto support these patterns as well? Would you update documentation for CAST to add these new patterns?

https://github.com/facebookincubator/velox/pull/5847

CC: @kagamiori

**Comment by marin-ma** [comment]
> @marin-ma Does Presto support these patterns as well? Would you update documentation for CAST to add these new patterns?
> 
> #5847
> 
> CC: @kagamiori

I'm not sure about prestosql. Neither do I have an test environment to check. @kagamiori Could you help to check on prestosql?

**Comment by marin-ma** [comment]
I just got a test environment for prestosql. Only `SELECT cast('YYYY-MM-DD' as date)` pass and all other patterns fails. Looks like prestosql only support one pattern 'YYYY-MM-DD' @kagamiori

**Comment by marin-ma** [comment]
> @marin-ma Thank you for checking. Sounds like Presto and Spark behavior for cast(varchar as date) is different. Since, CastExpr is shared between the two, we would need to introduce a configuration property to control the behavior. Search for queryConfig.isCastToIntByTruncate() to see some examples.

@mbasmanova Updated & PTAL. Thanks!

File: velox/type/TimestampConversion.cpp [hunk 1]
```diff
@@ -105,6 +105,8 @@ constexpr int32_t kCumulativeYearDays[] = {
 
 namespace {
 
+enum class ParseMode { STRICT, NON_STRICT, STANDARD_CAST, NON_STANDARD_CAST };
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.cpp:108]: Please, review the coding guidelines and update accordingly.

File: velox/type/TimestampConversion.h [hunk 2]
```diff
@@ -81,6 +81,26 @@ inline int64_t fromDateString(const StringView& str) {
...
+///
+/// `[+-]YYYY*`
+/// `[+-]YYYY*-[M]M`
+/// `[+-]YYYY*-[M]M-[D]D`
+/// `[+-]YYYY*-[M]M-[D]D `
+/// `[+-]YYYY*-[M]M-[D]D *`
+/// `[+-]YYYY*-[M]M-[D]DT*`
+///
+/// Throws VeloxUserError if the format or date is invalid.
+int32_t castFromDateString(const char* buf, size_t len, bool isNonStandardCast);
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.h:96]: It would be nice to extract this change into a separate PR and add tests.

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to review?

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to review this PR? This one is a follow-up of https://github.com/facebookincubator/velox/pull/5994 Thanks!

**Comment by marin-ma** [comment]
@mbasmanova Could you help to review again?

File: velox/expression/tests/CastExprTest.cpp [hunk 3]
```diff
@@ -651,6 +658,68 @@ TEST_F(CastExprTest, invalidDate) {
       "date", {"2012-Oct-23"}, {0}, true, false, VARCHAR(), DATE());
 }
 
+TEST_F(CastExprTest, stringToDate) {
```
Comment (bikramSingh91) [inline velox/expression/tests/CastExprTest.cpp:661]: nit: can you move these tests to TEST_F(CastExprTest, date)

File: velox/expression/tests/CastExprTest.cpp [hunk 4]
```diff
@@ -651,6 +658,68 @@ TEST_F(CastExprTest, invalidDate) {
...
+       "2015-03-18T",
+       "2015-03-18T123123"},
+      {3789391, 16436, 16495, 16512, 16512, 16512, 16512},
+      false,
+      false,
+      VARCHAR(),
+      DATE());
+}
+
+TEST_F(CastExprTest, stringToDateInvalid) {
```
Comment (bikramSingh91) [inline velox/expression/tests/CastExprTest.cpp:697]: nit: can you move these tests to TEST_F(CastExprTest, invalidDate)

File: velox/docs/configs.rst [hunk 5]
```diff
@@ -109,6 +109,11 @@ Expression Evaluation Configuration
      - bool
      - false
      - This flags forces the cast from float/double to integer to be performed by truncating the decimal part instead of rounding.
+   * - cast_string_to_date_non_standard
+     - bool
+     - false
+     - This flags allows the cast from string to date accept patterns other than the "[+-](YYYY-MM-DD)" format.
+       Valid patterns include (YYYY, YYYY-MM, YYYY-MM-DD), and any patterns prefixed with [+-]".
```
Comment (bikramSingh91) [inline velox/docs/configs.rst:116]: should we also mention other cases like
Digits after trailing spaces: "YYYY-MM-DD 123"
Trailing character 'T': "YYYY-MM-DDT"
Digits after 'T': "YYYY-MM-DDT123123"

File: velox/expression/tests/CastExprTest.cpp [hunk 6]
```diff
@@ -651,6 +658,68 @@ TEST_F(CastExprTest, invalidDate) {
       "date", {"2012-Oct-23"}, {0}, true, false, VARCHAR(), DATE());
 }
 
+TEST_F(CastExprTest, stringToDate) {
+  for (bool conf : {true, false}) {
+    setCastStringToDateNonStandard(conf);
+    testCast<std::string, int32_t>(
+        "date",
```
Comment (bikramSingh91) [inline velox/expression/tests/CastExprTest.cpp:665]: wondering if we should include a case for trailing (BC) for the case where it is valid

File: velox/core/QueryConfig.h [hunk 7]
```diff
@@ -80,6 +80,11 @@ class QueryConfig {
   // truncating the decimal part instead of rounding.
   static constexpr const char* kCastToIntByTruncate = "cast_to_int_by_truncate";
 
+  // This flags allows the cast from string to date accept patterns other than
+  // the "(+-)[YYYY-MM-DD]" format.
+  static constexpr const char* kCastStringToDateNonStandard =
```
Comment (bikramSingh91) [inline velox/core/QueryConfig.h:85]: nit: can you also mention this newly added query config in the PR description that enables this behavior.

File: velox/docs/configs.rst [hunk 8]
```diff
@@ -109,6 +109,11 @@ Expression Evaluation Configuration
      - bool
      - false
      - This flags forces the cast from float/double to integer to be performed by truncating the decimal part instead of rounding.
+   * - cast_string_to_date_non_standard
```
Comment (mbasmanova) [inline velox/docs/configs.rst:112]: Let's also document configs that alter cast behavior in cast documentation: https://facebookincubator.github.io/velox/functions/presto/conversion.html

Comment (marin-ma) [reply velox/docs/configs.rst:112]: @mbasmanova "non_standard" here means non ISO8601 standard. Or change it to `cast_string_to_date_iso8601` and make the default value true?

File: velox/expression/tests/CastExprTest.cpp (hunk 6 above)
Comment (marin-ma) [reply velox/expression/tests/CastExprTest.cpp:618]: Added. It's invalid in presto but valid in spark.

Comment (marin-ma) [reply velox/expression/tests/CastExprTest.cpp:618]: @bikramSingh91 My previous understanding of trailing (BC) was wrong. The trailing (BC) starts with an additional space. So `cast("2015-03-18(BC)" as date` returns NULL in spark but `cast("2015-03-18 (BC)" as date` returns `2015-03-18`

**Comment by marin-ma** [comment]
@mbasmanova @bikramSingh91 Could you help to review again? Thanks!

File: velox/docs/configs.rst [hunk 9]
```diff
@@ -108,7 +108,13 @@ Expression Evaluation Configuration
    * - cast_to_int_by_truncate
      - bool
      - false
-     - This flags forces the cast from float/double to integer to be performed by truncating the decimal part instead of rounding.
+     - This flag forces the cast from float/double to integer to be performed by truncating the decimal part instead of rounding.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:111]: drop 'This flag' and use active verbs for ease of reading

```
If set, cast from float or double to integer truncates the decimal part., otherwise rounds.
```

File: velox/docs/configs.rst [hunk 10]
```diff
@@ -108,7 +108,13 @@ Expression Evaluation Configuration
...
+   * - cast_string_to_date_iso_8601
+     - bool
+     - true
+     - This flag sets the cast from string to date accepts patterns with the "[+-](YYYY-MM-DD)" format (ISO 8601).
```
Comment (mbasmanova) [inline velox/docs/configs.rst:115]:
```
If set, cast from string to date allows only ISO 8601 formatted strings: [+-](YYYY-MM-DD). Otherwise, allows all patterns supported by Spark. <TBD include a full list of pattern and a link to Spark's documentation if available>
```

File: velox/docs/functions/presto/conversion.rst [hunk 11]
```diff
@@ -455,7 +455,10 @@ From strings
 ^^^^^^^^^^^^
 
 Casting from a string to date is allowed if the string represents a date in the
```
Comment (mbasmanova) [inline velox/docs/functions/presto/conversion.rst:457]:
```
By default, only ISO 8601 strings are supported: [+-](YYYY-MM-DD).

If cast_string_to_date_ios_8601 is set, all Spark supported patterns are allowed. See <link to documentation for cast_string_to_date_ios_8601> for the full list of supported patterns.
```

Please, add examples for the case when cast_string_to_date_ios_8601 is true.

File: velox/expression/CastExpr.cpp [hunk 12]
```diff
@@ -100,10 +100,13 @@ VectorPtr CastExpr::castToDate(
   switch (fromType->kind()) {
     case TypeKind::VARCHAR: {
       auto* inputVector = input.as<SimpleVector<StringView>>();
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      auto isStandardCast = queryConfig.isCastStringToDateIso8601();
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:104]: nit: perhaps, use more specific name isIso8601

File: velox/expression/tests/CastExprTest.cpp [hunk 13]
```diff
@@ -606,49 +612,101 @@ TEST_F(CastExprTest, timestampAdjustToTimezoneInvalid) {
...
-      60577,
-      -5,
-      -57604,
-      -18262,
-      std::nullopt,
-  };
-
-  testCast<std::string, int32_t>(
-      "date", input, result, false, false, VARCHAR(), DATE());
+  for (bool conf : {true, false}) {
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:615]: do not abbereviate: conf -> isIso8601

File: velox/type/TimestampConversion.cpp [hunk 14]
```diff
@@ -579,26 +579,24 @@ int64_t fromDateString(const char* str, size_t len) {
...
+  auto mode =
+      isStandardCast ? ParseMode::kStandardCast : ParseMode::kNonStandardCast;
   if (!tryParseDateString(str, len, pos, daysSinceEpoch, mode)) {
-    if (isNonStandardCast) {
+    if (isStandardCast) {
       VELOX_USER_FAIL(
           "Unable to parse date value: \"{}\"."
-          "Valid date string patterns include "
-          "(YYYY, YYYY-MM, YYYY-MM-DD), and any pattern prefixed with [+-]",
+          "Valid date string pattern is (YYYY-MM-DD), "
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.cpp:592]: Exceptions are expensive and cause performance issues when queries use try_cast and have many failing rows. It would be nice to re-write this logic to not use exceptions.

File: velox/type/TimestampConversion.cpp [hunk 15]
```diff
@@ -579,26 +579,24 @@ int64_t fromDateString(const char* str, size_t len) {
...
+          "and can be prefixed with [+-]",
           std::string(str, len));
-
     } else {
       VELOX_USER_FAIL(
           "Unable to parse date value: \"{}\"."
-          "Valid date string pattern is (YYYY-MM-DD), "
-          "and can be prefixed with [+-]",
+          "Valid date string patterns include "
+          "(YYYY, YYYY-MM, YYYY-MM-DD), and any pattern prefixed with [+-]",
```
Comment (mbasmanova) [inline velox/type/TimestampConversion.cpp:599]: Is this the full list? What about BC suffix discussed above?

File: velox/type/tests/TimestampConversionTest.cpp [hunk 16]
```diff
@@ -110,42 +110,41 @@ TEST(DateTimeUtilTest, fromDateStrInvalid) {
...
+    EXPECT_THROW(castFromDateString("2015-031-8", standard), VeloxUserError);
   }
 
-  EXPECT_THROW(castFromDateString("12345", false), VeloxUserError);
-  EXPECT_THROW(castFromDateString("2015-03", false), VeloxUserError);
-  EXPECT_THROW(castFromDateString("2015-03-18 ", false), VeloxUserError);
-  EXPECT_THROW(castFromDateString("2015-03-18 123412", false), VeloxUserError);
-  EXPECT_THROW(castFromDateString("2015-03-18T", false), VeloxUserError);
-  EXPECT_THROW(castFromDateString("2015-03-18T123412", false), VeloxUserError);
+  EXPECT_THROW(castFromDateString("12345", true), VeloxUserError);
```
Comment (mbasmanova) [inline velox/type/tests/TimestampConversionTest.cpp:142]: use VELOX_ASSERT_THROW

**Review by mbasmanova — COMMENTED** [review]
Some further comments.

File: velox/docs/configs.rst [hunk 17]
```diff
@@ -116,11 +116,11 @@ Expression Evaluation Configuration
...
-         * ``[+-]yyyy*``
-         * ``[+-]yyyy*-[m]m``
-         * ``[+-]yyyy*-[m]m-[d]d``
-         * ``[+-]yyyy*-[m]m-[d]d *``
-         * ``[+-]yyyy*-[m]m-[d]dT*``
+       ``[+-]yyyy*``,
+       ``[+-]yyyy*-[m]m``,
+       ``[+-]yyyy*-[m]m-[d]d``,
+       ``[+-]yyyy*-[m]m-[d]d *``,
+       ``[+-]yyyy*-[m]m-[d]dT*``
```
Comment (mbasmanova) [inline velox/docs/configs.rst:123]: Showing as a bullet list is probably more readable then comma-separated list. 

Also, should we include (BC) patterns?

Comment (marin-ma) [reply velox/docs/configs.rst:123]: @mbasmanova  (BC) pattern is a special case of `[+-]yyyy*-[m]m-[d]d *`

Comment (mbasmanova) [reply velox/docs/configs.rst:123]: Interesting. I guess I do not understand what "*" in the pattern means. Would you clarify? Let's also update the docs to make sure other readers are not misled.

Comment (marin-ma) [reply velox/docs/configs.rst:123]: The documents are copied from Spark's comment https://github.com/apache/spark/blob/3e5203c64c06cc8a8560dfa0fb6f52e74589b583/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L286-L298

I think the meaning of `*` in Spark's comment is any numbers, like `yyyy*`. But Spark doesn't check whether the trailing characters after 'T' or space are numbers, so in fact any characters are allowed.

Comment (mbasmanova) [reply velox/docs/configs.rst:123]: > Spark doesn't check whether the trailing characters after 'T' or space are numbers, so in fact any characters are allowed.

I see... Let's clarify this in the doc and provide examples. This is non-obvious.

File: velox/docs/functions/presto/conversion.rst [hunk 18]
```diff
@@ -454,20 +454,53 @@ Cast to Date
...
-Invalid example
+**cast_string_to_date_is_iso_8601=false**
+
+::
+
+  SELECT cast('1970' as date); -- 1970-01-01
+  SELECT cast('1970-01' as date); -- 1970-01-01
+  SELECT cast('1970-01-01' as date); -- 1970-01-01
+  SELECT cast('1970-01-01T123' as date); -- 1970-01-01
+  SELECT cast('1970-01-01 (BC)' as date); -- 1970-01-01
```
Comment (marin-ma) [inline velox/docs/functions/presto/conversion.rst:481]: @mbasmanova Example for trailing characters other than numbers has been added here. Do we need still to show examples in https://facebookincubator.github.io/velox/configs.html#generic-configuration?

Comment (mbasmanova) [reply velox/docs/functions/presto/conversion.rst:481]: It would be nice to add these to the other doc as well. Makes it easier for the readers.

**Comment by marin-ma** [comment]
@mbasmanova Could you help review again? Thanks!

**Review by mbasmanova — COMMENTED** [review]
@marin-ma I'll defer to Bikram to finish the review and help merge.

File: velox/expression/CastExpr.cpp [hunk 19]
```diff
@@ -100,16 +121,21 @@ VectorPtr CastExpr::castToDate(
...
-          resultFlatVector->set(row, DATE()->toDays(inputString));
-        } catch (const VeloxUserError& ue) {
-          VELOX_USER_FAIL(
-              makeErrorMessage(input, row, DATE()) + " " + ue.message());
-        } catch (const std::exception& e) {
-          VELOX_USER_FAIL(
-              makeErrorMessage(input, row, DATE()) + " " + e.what());
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      auto isStandardCast = queryConfig.isIso8601();
+      context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:126]: Should we use `applyToSelectedNoThrowLocal`? If castFromDateString throws exception, the setNullInResultAtError flag will be ignored.

File: velox/core/QueryConfig.h [hunk 20]
```diff
@@ -76,10 +76,26 @@ class QueryConfig {
   static constexpr const char* kCastMatchStructByName =
       "cast_match_struct_by_name";
 
-  // This flags forces the cast from float/double to integer to be performed by
-  // truncating the decimal part instead of rounding.
+  // If set, cast from float or double to integer truncates the decimal part,
+  // otherwise rounds.
   static constexpr const char* kCastToIntByTruncate = "cast_to_int_by_truncate";
 
+  // If set, cast from string to date allows only ISO 8601 formatted strings:
```
Comment (jinchengchenghh) [inline velox/core/QueryConfig.h:83]: /// for the multiple line in header file

File: velox/expression/tests/CastExprTest.cpp [hunk 21]
```diff
@@ -615,49 +621,101 @@ TEST_F(CastExprTest, timestampAdjustToTimezoneInvalid) {
...
+         -719893,
+         std::nullopt},
+        false,
+        false,
+        VARCHAR(),
+        DATE());
+  }
 
-  setCastIntByTruncate(true);
+  // Valid cases for cast_string_to_date_iso_8601=false
```
Comment (jinchengchenghh) [inline velox/expression/tests/CastExprTest.cpp:658]: dot in the end of line

File: velox/type/tests/TimestampConversionTest.cpp [hunk 22]
```diff
@@ -110,42 +110,43 @@ TEST(DateTimeUtilTest, fromDateStrInvalid) {
...
-    EXPECT_EQ(0, castFromDateString("1970-01-01", nonStandard));
-    EXPECT_EQ(3789742, castFromDateString("12345-12-18", nonStandard));
+  for (bool standard : {true, false}) {
+    EXPECT_EQ(0, castFromDateString("1970-01-01", standard));
+    EXPECT_EQ(3789742, castFromDateString("12345-12-18", standard));
 
-    EXPECT_EQ(1, castFromDateString("+1970-1-2", nonStandard));
-    EXPECT_EQ(1, castFromDateString("+1970-01-2", nonStandard));
-    EXPECT_EQ(1, castFromDateString("+1970-1-02", nonStandard));
+    EXPECT_EQ(1, castFromDateString("+1970-1-2", standard));
```
Comment (jinchengchenghh) [inline velox/type/tests/TimestampConversionTest.cpp:117]: Presto and spark also support this format, can you test this one?
```
presto:default> SELECT cast('1970-01-01 ' as date);
   _col0
------------
 1970-01-01
(1 row)

spark-sql> SELECT cast('1970-01-01 ' as date);
1970-01-01
Time taken: 2.703 seconds, Fetched 1 row(s)
```

Comment (marin-ma) [reply velox/type/tests/TimestampConversionTest.cpp:117]: @jinchengchenghh Thanks. This pattern is not supported in this PR. I'll fix it. Could you help to check whether prefix spaces are valid in presto?

`select cast(' 1970-01-01' as date);`

Comment (jinchengchenghh) [reply velox/type/tests/TimestampConversionTest.cpp:118]: presto:default> select cast(' 1970-01-01' as date);
   _col0
------------
 1970-01-01
(1 row)

Query 20230828_085646_00002_sqyi8, FINISHED, 1 node

**Comment by jinchengchenghh** [comment]
Can you also add the test for error message by `VELOX_ASSERT_THROW` though it is different with Presto error message?
presto:default> SELECT cast('2012-Oct-23' as date);
Query 20230828_101619_00004_sqyi8 failed: Value cannot be cast to date: 2012-Oct-23

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help review again? Thanks!

File: velox/expression/CastExpr.cpp [hunk 23]
```diff
@@ -103,10 +124,22 @@ VectorPtr CastExpr::castToDate(
...
       applyToSelectedNoThrowLocal(context, rows, castResult, [&](int row) {
         try {
           auto inputString = inputVector->valueAt(row);
-          resultFlatVector->set(row, DATE()->toDays(inputString));
+          auto date = util::castFromDateString(inputString, isStandardCast);
+          if (date.has_value()) {
+            resultFlatVector->set(row, *date);
+          } else {
+            if (setNullInResultAtError()) {
+              resultFlatVector->setNull(row, true);
```
Comment (bikramSingh91) [inline velox/expression/CastExpr.cpp:137]: we dont need to handle this case here since applyToSelectedNoThrowLocal() will take care of setting the errors for the respective rows and there is common logic to handle setting nulls on error.

Comment (bikramSingh91) [reply velox/expression/CastExpr.cpp:137]: Apart from the case where nullptr is returned, there are also a case in castFromDateString() that return a velox user error (daysSinceEpochFromDate) still which needs to be handled in a consistent way. The previous patchset that returned a VELOX_USER_ERROR should suffice

**Comment by marin-ma** [comment]
@bikramSingh91 Addressed the comments. Could you help to review again? Thanks!

File: velox/type/TimestampConversion.cpp [hunk 24]
```diff
@@ -579,26 +588,24 @@ int64_t fromDateString(const char* str, size_t len) {
...
```
Comment (bikramSingh91) [inline velox/type/TimestampConversion.cpp:608]: nit: Can you include the full list of supported patterns here like you mentioned in configs.rst

**Review by bikramSingh91 — APPROVED** [review]
Looks good, thanks for working on this. just a small nit left about the error message and a rebase. I can then pull in changes to run tests and start the merge process.

**Comment by bikramSingh91** [comment]
@marin-ma Can you please take a look at the failure in `castFromDateStringInvalid` that failing some of the circleCI precommits?

**Comment by marin-ma** [comment]
> @marin-ma Can you please take a look at the failure in `castFromDateStringInvalid` that failing some of the circleCI precommits?

PTAL. Thanks!

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to check the error message of Linter? I can't see the errors.

File: velox/type/tests/TimestampConversionTest.cpp [hunk 25]
```diff
@@ -15,6 +15,7 @@
  */
 
 #include "velox/type/TimestampConversion.h"
+#include <common/base/tests/GTestUtils.h>
```
Comment (bikramSingh91) [inline velox/type/tests/TimestampConversionTest.cpp:18]: nit: can you  please update this to use 
`#include "velox/common/base/tests/GTestUtils.h"`
and add it under L21 instead

File: velox/type/tests/TimestampConversionTest.cpp [hunk 26]
```diff
@@ -15,6 +15,7 @@
...
 #include <gmock/gmock.h>
 #include <gtest/gtest.h>
```
Comment (jinchengchenghh) [inline velox/type/tests/TimestampConversionTest.cpp:20]: Include `#include "velox/common/base/tests/GTestUtils.h"` we can remove `#include <gtest/gtest.h>`

File: velox/type/TimestampConversion.cpp [hunk 27]
```diff
@@ -268,8 +268,16 @@ bool tryParseDateString(
...
   if (mode == ParseMode::kStandardCast) {
+    // Skip trailing spaces.
+    while (pos < len && characterIsSpace(buf[pos])) {
+      pos++;
+    }
+    // Check position. if end was not reached, non-space chars remaining.
+    if (pos < len) {
+      return false;
+    }
+
```
Comment (bikramSingh91) [inline velox/type/TimestampConversion.cpp:280]: @marin-ma Thanks for making all the changes till now. I have a last request, can you please move this part of the change out in a separate PR? This unfortunately, is a change in behavior of existing `cast` and can result in issues for existing workloads that dont expect this behavior. It would be good to separate it in case we need to roll back this behavior change while still keeping all the spark specific functionality you added here.

**Comment by bikramSingh91** [comment]
Thank you for moving out the changes to standard mode. I'll import these updates and will proceed with merging.

**Comment by marin-ma** [comment]
@bikramSingh91 Removed the changes in the latest commit. Could you help review again? Thanks!

**Comment by marin-ma** [comment]
@bikramSingh91 Could you help to review again? Thanks!

**Comment by bikramSingh91** [comment]
@marin-ma I was on PTO for that last week, will take this up today with high priority
